File: flighttracker/analyze-tropical-flights.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

async function analyzeTropicalFlights() {
  console.log('📊 ANALYZING TROPICAL FLIGHTS-ONLY RESULTS');
  console.log('🎯 Target: Family of 4 (2A + 2C) - Under £1500 - 5 Days August 2025');
  console.log('✈️ FLIGHTS ONLY - No packages\n');

  // Get all tropical flights result files
  const dataDir = path.join(__dirname, 'data');
  const files = fs.readdirSync(dataDir).filter(file => 
    file.startsWith('flight-history-tropical-flights-') && file.endsWith('.json')
  );

  console.log(`📁 Found ${files.length} flight search result files`);

  const allResults = [];
  const destinationStats = {};
  const airportStats = {};

  // Tropical destinations mapping
  const tropicalDestinations = {
    'TFS': { name: 'Tenerife', country: 'Spain' },
    'LPA': { name: 'Gran Canaria', country: 'Spain' },
    'ACE': { name: 'Lanzarote', country: 'Spain' },
    'AYT': { name: 'Antalya', country: 'Turkey' },
    'IST': { name: 'Istanbul', country: 'Turkey' },
    'LCA': { name: 'Cyprus', country: 'Cyprus' },
    'ATH': { name: 'Athens', country: 'Greece' },
    'HER': { name: 'Crete', country: 'Greece' },
    'MLA': { name: 'Malta', country: 'Malta' },
    'CMN': { name: 'Casablanca', country: 'Morocco' },
    'HRG': { name: 'Hurghada', country: 'Egypt' }
  };

  const ukAirports = {
    'LHR': 'London Heathrow',
    'LGW': 'London Gatwick', 
    'STN': 'London Stansted',
    'MAN': 'Manchester'
  };

  // Process each file
  for (const file of files) {
    try {
      const filePath = path.join(dataDir, file);
      const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
      
      if (data.bestPrice && data.bestPrice.result) {
        const result = data.bestPrice.result;
        
        // Extract route info from filename
        const filenameParts = file.replace('flight-history-tropical-flights-', '').replace('.json', '').split('-');
        let destination = 'unknown';
        let origin = 'unknown';
        let dates = 'unknown';
        
        // Parse filename to extract destination and origin
        if (filenameParts.length >= 3) {
          // Find destination (should be one of our tropical destinations)
          for (let i = 0; i < filenameParts.length - 1; i++) {
            const possibleDest = filenameParts[i].toUpperCase();
            if (Object.keys(tropicalDestinations).includes(possibleDest)) {
              destination = possibleDest;
              if (i + 1 < filenameParts.length) {
                origin = filenameParts[i + 1].toUpperCase();
              }
              if (i + 2 < filenameParts.length) {
                dates = filenameParts.slice(i + 2).join('-');
              }
              break;
            }
          }
          
          // Alternative parsing - look for known airport codes
          if (destination === 'unknown') {
            for (const part of filenameParts) {
              if (Object.keys(tropicalDestinations).some(dest => 
                part.toLowerCase().includes(dest.toLowerCase()) || 
                tropicalDestinations[dest].name.toLowerCase().includes(part.toLowerCase())
              )) {
                // Found destination match
                for (const [code, info] of Object.entries(tropicalDestinations)) {
                  if (part.toLowerCase().includes(info.name.toLowerCase()) || 
                      part.toLowerCase().includes(code.toLowerCase())) {
                    destination = code;
                    break;
                  }
                }
                break;
              }
            }
          }
          
          // Find UK airport
          if (origin === 'unknown') {
            for (const part of filenameParts) {
              const upperPart = part.toUpperCase();
              if (Object.keys(ukAirports).includes(upperPart)) {
                origin = upperPart;
                break;
              }
            }
          }
        }

        const destInfo = tropicalDestinations[destination] || { name: destination, country: 'Unknown' };
        const originInfo = ukAirports[origin] || origin;
        
        const flightResult = {
          destination: destination,
          destinationName: destInfo.name,
          destinationCountry: destInfo.country,
          origin: origin,
          originName: originInfo,
          price: result.price,
          currency: result.currency || 'GBP',
          airline: result.airline,
          duration: result.duration,
          stops: result.stops || 0,
          bookingUrl: result.deepLink,
          route: `${originInfo} → ${destInfo.name}`,
          passengers: '2A + 2C',
          dates: dates,
          filename: file
        };

        allResults.push(flightResult);

        // Update destination stats
        if (!destinationStats[destination]) {
          destinationStats[destination] = {
            name: destInfo.name,
            country: destInfo.country,
            count: 0,
            minPrice: Infinity,
            maxPrice: 0,
            totalPrice: 0,
            routes: []
          };
        }
        destinationStats[destination].count++;
        destinationStats[destination].minPrice = Math.min(destinationStats[destination].minPrice, result.price);
        destinationStats[destination].maxPrice = Math.max(destinationStats[destination].maxPrice, result.price);
        destinationStats[destination].totalPrice += result.price;
        destinationStats[destination].routes.push(flightResult);

        // Update airport stats
        if (!airportStats[origin]) {
          airportStats[origin] = {
            name: originInfo,
            count: 0,
            minPrice: Infinity,
            maxPrice: 0,
            totalPrice: 0
          };
        }
        airportStats[origin].count++;
        airportStats[origin].minPrice = Math.min(airportStats[origin].minPrice, result.price);
        airportStats[origin].maxPrice = Math.max(airportStats[origin].maxPrice, result.price);
        airportStats[origin].totalPrice += result.price;
      }
    } catch (error) {
      console.error(`❌ Error processing ${file}:`, error.message);
    }
  }

  // Sort results by price
  allResults.sort((a, b) => a.price - b.price);

  // Filter results under £1500
  const affordableResults = allResults.filter(r => r.price <= 1500);
  const expensiveResults = allResults.filter(r => r.price > 1500).slice(0, 10);

  console.log('\n' + '='.repeat(80));
  console.log('🏆 TROPICAL FLIGHTS-ONLY ANALYSIS RESULTS');
  console.log('🎯 Family of 4 (2A + 2C) - 5 Days August 2025');
  console.log('='.repeat(80));

  console.log(`📊 Total searches analyzed: ${allResults.length}`);
  console.log(`💰 Budget target: £1500 for family of 4`);

  if (affordableResults.length === 0) {
    console.log('\n❌ NO FLIGHTS FOUND UNDER £1500 FOR FAMILY OF 4');
    console.log('💡 All tropical destinations exceeded the £1500 budget');
    
    console.log('\n📊 CLOSEST OPTIONS (Above £1500):');
    console.log('-'.repeat(60));
    expensiveResults.forEach((result, index) => {
      const overBudget = result.price - 1500;
      console.log(`${(index + 1).toString().padStart(2)}. ${result.route}`);
      console.log(`    💰 £${result.price} total (${result.passengers})`);
      console.log(`    📈 £${overBudget} over £1500 budget`);
      console.log(`    ✈️  ${result.airline} (${result.stops} stops)`);
      console.log(`    📅 ${result.dates}`);
      console.log();
    });
  } else {
    console.log(`\n✅ Found ${affordableResults.length} affordable options under £1500!`);

    // Group by price ranges
    const priceRanges = {
      'Under £500': affordableResults.filter(r => r.price < 500),
      '£500-£750': affordableResults.filter(r => r.price >= 500 && r.price < 750),
      '£750-£1000': affordableResults.filter(r => r.price >= 750 && r.price < 1000),
      '£1000-£1250': affordableResults.filter(r => r.price >= 1000 && r.price < 1250),
      '£1250-£1500': affordableResults.filter(r => r.price >= 1250 && r.price <= 1500)
    };

    for (const [range, results] of Object.entries(priceRanges)) {
      if (results.length > 0) {
        console.log(`\n💰 ${range} (${results.length} options):`);
        results.slice(0, 5).forEach((result, index) => {
          console.log(`  ${index + 1}. ${result.route}`);
          console.log(`     💰 £${result.price} total (${result.passengers})`);
          console.log(`     ✈️  ${result.airline} (${result.stops} stops)`);
          console.log(`     📅 ${result.dates}`);
          console.log(`     🔗 ${result.bookingUrl}`);
          console.log();
        });
        if (results.length > 5) {
          console.log(`     ... and ${results.length - 5} more options\n`);
        }
      }
    }

    console.log('\n🏆 TOP 10 BEST DEALS (Under £1500):');
    console.log('-'.repeat(60));
    affordableResults.slice(0, 10).forEach((result, index) => {
      const savings = 1500 - result.price;
      console.log(`${(index + 1).toString().padStart(2)}. ${result.route}`);
      console.log(`    💰 £${result.price} total (${result.passengers})`);
      console.log(`    💚 £${savings} under budget`);
      console.log(`    ✈️  ${result.airline} (${result.stops} stops)`);
      console.log(`    📅 ${result.dates}`);
      console.log(`    🔗 ${result.bookingUrl}`);
      console.log();
    });
  }

  // Destination analysis
  console.log('\n📊 DESTINATION ANALYSIS:');
  console.log('-'.repeat(50));
  Object.entries(destinationStats)
    .sort(([,a], [,b]) => a.minPrice - b.minPrice)
    .forEach(([code, stats]) => {
      const avgPrice = Math.round(stats.totalPrice / stats.count);
      const affordableCount = stats.routes.filter(r => r.price <= 1500).length;
      console.log(`${stats.name}, ${stats.country} (${code}):`);
      console.log(`  📊 ${stats.count} routes | From £${stats.minPrice} | Avg £${avgPrice}`);
      console.log(`  ✅ ${affordableCount} under £1500`);
      console.log();
    });

  // UK airport analysis
  console.log('\n🛫 UK DEPARTURE AIRPORT ANALYSIS:');
  console.log('-'.repeat(50));
  Object.entries(airportStats)
    .sort(([,a], [,b]) => a.minPrice - b.minPrice)
    .forEach(([code, stats]) => {
      const avgPrice = Math.round(stats.totalPrice / stats.count);
      console.log(`${stats.name} (${code}):`);
      console.log(`  📊 ${stats.count} routes | From £${stats.minPrice} | Avg £${avgPrice}`);
      console.log();
    });

  // Summary recommendations
  console.log('\n💡 RECOMMENDATIONS:');
  console.log('-'.repeat(30));
  if (affordableResults.length === 0) {
    console.log('• Consider increasing budget above £1500');
    console.log('• Look at different travel dates in August 2025');
    console.log('• Consider 3-4 day trips instead of 5 days');
    console.log('• Check for last-minute deals closer to travel dates');
    
    // Find the cheapest option
    if (allResults.length > 0) {
      const cheapest = allResults[0];
      console.log(`• Cheapest option found: ${cheapest.route} at £${cheapest.price}`);
    }
  } else {
    const bestDeal = affordableResults[0];
    console.log(`• Best deal: ${bestDeal.route} at £${bestDeal.price}`);
    console.log(`• You have ${affordableResults.length} affordable options to choose from`);
    console.log('• Book early for better prices and availability');
    console.log('• Consider travel insurance for family trips');
  }

  console.log('\n' + '='.repeat(80));
  console.log('✅ Tropical flights-only analysis complete!');
  console.log(`📊 Analyzed ${allResults.length} flight combinations`);
  console.log(`🎯 Found ${affordableResults.length} options under £1500 budget`);
  console.log('🏝️ Focus: 5-day tropical trips in August 2025');
  console.log('✈️ FLIGHTS ONLY - No accommodation included');
  console.log('👨‍👩‍👧‍👦 Family of 4 (2 Adults + 2 Children)');
  console.log('='.repeat(80));

  return {
    total: allResults.length,
    affordable: affordableResults.length,
    cheapest: allResults[0],
    bestAffordable: affordableResults[0] || null,
    destinations: Object.keys(destinationStats).length,
    airports: Object.keys(airportStats).length
  };
}

// Run the analysis
if (require.main === module) {
  analyzeTropicalFlights().catch(error => {
    console.error('❌ Analysis failed:', error.message);
    process.exit(1);
  });
}

module.exports = analyzeTropicalFlights;

File: flighttracker/analyze_tropical_flights.py
#!/usr/bin/env python3
import json
import os
import glob
from collections import defaultdict

def extract_flight_info(file_path):
    """Extract flight information from a JSON file."""
    try:
        with open(file_path, 'r') as f:
            data = json.load(f)
        
        # Parse filename to extract destination and departure airport
        filename = os.path.basename(file_path)
        base = filename.replace('flight-history-tropical-flights-', '').replace('.json', '')
        
        # Handle gran-canaria specially since it has a hyphen in the name
        if base.startswith('gran-canaria-'):
            destination = 'gran-canaria'
            remaining = base[13:]  # Remove 'gran-canaria-'
            parts = remaining.split('-')
            departure_airport = parts[0] if parts else None
        else:
            parts = base.split('-')
            if len(parts) >= 2:
                destination = parts[0]
                departure_airport = parts[1]
            else:
                destination = None
                departure_airport = None
        
        # Only proceed if we have valid destination and departure airport
        if destination and departure_airport:
            # Extract best price info
            if 'bestPrice' in data and data['bestPrice']:
                best_price = data['bestPrice']
                price = best_price.get('price')
                result = best_price.get('result', {})
                
                return {
                    'destination': destination,
                    'departure_airport': departure_airport,
                    'price': price,
                    'airline': result.get('airline'),
                    'stops': result.get('stops'),
                    'duration': result.get('duration'),  # in minutes
                    'currency': result.get('currency'),
                    'filename': filename
                }
    except Exception as e:
        print(f"Error processing {file_path}: {e}")
    
    return None

def format_duration(minutes):
    """Convert minutes to hours and minutes format."""
    if minutes is None:
        return "N/A"
    hours = minutes // 60
    mins = minutes % 60
    return f"{hours}h {mins}m"

def format_stops(stops):
    """Format stops information."""
    if stops is None:
        return "N/A"
    if stops == 0:
        return "Direct"
    elif stops == 1:
        return "1 stop"
    else:
        return f"{stops} stops"

def map_destination_codes():
    """Map destination codes to full names and airport codes."""
    return {
        'tenerife': ('Tenerife', 'TFS'),
        'gran-canaria': ('Gran Canaria', 'LPA'),
        'lanzarote': ('Lanzarote', 'ACE'),
        'antalya': ('Antalya', 'AYT'),
        'istanbul': ('Istanbul', 'IST'),
        'cyprus': ('Cyprus', 'LCA'),
        'athens': ('Athens', 'ATH'),
        'crete': ('Crete', 'HER'),
        'malta': ('Malta', 'MLA'),
        'casablanca': ('Casablanca', 'CMN'),
        'hurghada': ('Hurghada', 'HRG')
    }

def map_airport_codes():
    """Map airport codes to full names."""
    return {
        'lgw': 'London Gatwick (LGW)',
        'lhr': 'London Heathrow (LHR)',
        'stn': 'London Stansted (STN)',
        'man': 'Manchester (MAN)'
    }

def main():
    # Find all tropical flight files
    data_dir = "/Users/sivaprakasamsivaprakasam/agents/flighttracker/data"
    pattern = os.path.join(data_dir, "flight-history-tropical-flights-*.json")
    files = glob.glob(pattern)
    
    print(f"Found {len(files)} tropical flight files to analyze")
    
    # Process all files
    all_flights = []
    for file_path in files:
        flight_info = extract_flight_info(file_path)
        if flight_info and flight_info['price'] is not None:
            all_flights.append(flight_info)
    
    print(f"Successfully processed {len(all_flights)} flight records")
    
    # Group by destination
    destinations = defaultdict(list)
    for flight in all_flights:
        destinations[flight['destination']].append(flight)
    
    # Sort each destination by price and get top 3
    destination_names = map_destination_codes()
    airport_names = map_airport_codes()
    
    print("\n" + "="*100)
    print("TOP 3 BEST VALUE TROPICAL FLIGHT DEALS - FAMILY OF 4")
    print("="*100)
    
    for dest_code in sorted(destinations.keys()):
        if dest_code in destination_names:
            dest_name, dest_airport_code = destination_names[dest_code]
            flights = sorted(destinations[dest_code], key=lambda x: x['price'])
            
            print(f"\n🏝️  {dest_name.upper()} ({dest_airport_code})")
            print("-" * 80)
            
            # Show top 3 best deals
            top_flights = flights[:3]
            for i, flight in enumerate(top_flights, 1):
                departure_airport = airport_names.get(flight['departure_airport'], flight['departure_airport'].upper())
                
                print(f"\n{i}. £{flight['price']:,} - {departure_airport}")
                print(f"   ✈️  Airline: {flight['airline']}")
                print(f"   🛣️  Route: {format_stops(flight['stops'])}")
                print(f"   ⏱️  Duration: {format_duration(flight['duration'])}")
            
            # Show summary stats
            all_prices = [f['price'] for f in flights]
            avg_price = sum(all_prices) / len(all_prices)
            print(f"\n   📊 Price Range: £{min(all_prices):,} - £{max(all_prices):,} (Avg: £{avg_price:.0f})")
            print(f"   🔍 Searched: {len(flights)} combinations")

    # Overall summary
    print(f"\n" + "="*100)
    print("SUMMARY")
    print("="*100)
    
    # Find absolute cheapest across all destinations
    all_flights_sorted = sorted(all_flights, key=lambda x: x['price'])
    cheapest_overall = all_flights_sorted[:5]
    
    print(f"\n🏆 TOP 5 CHEAPEST TROPICAL DESTINATIONS:")
    for i, flight in enumerate(cheapest_overall, 1):
        dest_name, dest_code = destination_names.get(flight['destination'], (flight['destination'], 'N/A'))
        departure_airport = airport_names.get(flight['departure_airport'], flight['departure_airport'].upper())
        
        print(f"{i}. £{flight['price']:,} - {dest_name} from {departure_airport}")
        print(f"    {flight['airline']} • {format_stops(flight['stops'])} • {format_duration(flight['duration'])}")
    
    # Price analysis by departure airport
    print(f"\n📍 BEST PRICES BY DEPARTURE AIRPORT:")
    airport_best = defaultdict(list)
    for flight in all_flights:
        airport_best[flight['departure_airport']].append(flight['price'])
    
    for airport_code in sorted(airport_best.keys()):
        prices = airport_best[airport_code]
        airport_name = airport_names.get(airport_code, airport_code.upper())
        print(f"   {airport_name}: £{min(prices):,} - £{max(prices):,} (Avg: £{sum(prices)/len(prices):.0f})")
    
    print(f"\n💰 Family of 4 flights under £2000: {len([f for f in all_flights if f['price'] < 2000])}/{len(all_flights)}")

if __name__ == "__main__":
    main()